keanu-project/src/main/java/io/improbable/keanu/network/ProtobufWriter.java

```diff
@@ -0,0 +1,101 @@
+package io.improbable.keanu.network;
+
+import com.google.common.primitives.Longs;
+import io.improbable.keanu.KeanuSavedBayesNet;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Map;
+
+public class ProtobufWriter implements NetworkWriter {
+    private final BayesianNetwork net;
+    private KeanuSavedBayesNet.BayesianNetwork.Builder bayesNetBuilder = null;
+
+    public ProtobufWriter(BayesianNetwork net) {
+        this.net = net;
+    }
+
+    public void save(OutputStream output, boolean saveValues) throws IOException {
+        bayesNetBuilder = KeanuSavedBayesNet.BayesianNetwork.newBuilder();
+
+        net.save(this);
+
+        if (saveValues) {
+            net.saveValues(this);
+        }
+
+        bayesNetBuilder.build().writeTo(output);
+        bayesNetBuilder = null;
```

Comment:
Why does the builder need to be stored as an instance variable if it is created and destroyed within this method?

Reply:
Because there are lots of callbacks in to this object to add each Vertex/Value as you go along.  We'd otherwise need to pass that through to BayesNet/each vertex - removing the benefit of removing all the Protobuf related stuff from the Vertex object?